### ISSUE TITLE
Allowing pinner to pin content from within a collaboration

### ISSUE THREAD

**Issue by victorb** [ipfs-shipyard/peer-star-app#146]
Not sure about this idea but sharing it out loud to get feedback if valuable or not. 

If you have a collaboration for sharing information in some app, you could also share other objects within that collaboration. Image a feed of images for example, state could look like this:

```
[{
  title: 'Look at my fancy image!',
  image: {
    '/': 'QmAsd'
  }
}]
```

If the pinner sees this, it'll persist the state itself but it won't be pinning the content within the collaboration (yet, AFAIK). 

[`replicateOnly`](https://github.com/ipfs-shipyard/peer-star-app/blob/master/src/pinner/index.js#L117) seems to be activate, which I understand to mean you can't even call `shared.value()` right now. But even with that changed, it will not pin the content within.

## Solution

I'm unsure of a exact solution. I think allowing users to specify some JQ-like string as a configuration when starting the pinner can solve this easy usecase but maybe not more complicated ones. With that, the user could do `pinner --pin-within [].image["/"]` and it would use that to pin specific hashes inside the collaboration.

**Comment by pgte** [comment]
@victorb are you talking about public static content pinning? Or private?

**Comment by victorb** [comment]
Yeah, I'm talking about content that is defined within the collaboration but it's defined as IPLD object (example above) in public collaborations (otherwise the pinner needs the same keys, gets more complicated).

For example, sharing images in a feed. You don't want to have to load the image itself when loading the collaboration, but rather load all the index of the images, and then each image after each other. So you embed a multihash (or even better, use DAG objects) so when the pinner sees this, it should *somehow* (to be defined) pin that content + the index.

**Comment by pgte** [comment]
@victorb if the content is public, could we use the IPFS pinning interface instead?

**Comment by victorb** [comment]
@pgte yeah, under the hood, I imagine the pinner will use the pinning interface, but it'll be handy if people can deploy one pinner to have their content + state replicated rather than having to deploy two services + make apps pin the content via another daemon AND pinning the state via the pinner.

**Comment by pgte** [comment]
@victorb I see, makes sense.

Implementation idea: instead of the peer having to contact the pinner directly, we could make the pinning collaborative: the files to pin are kept inside a dir (a CRDT), and the root hash is then what the pinner pins.

**Comment by satazor** [comment]
I'm shimming in because I currently have this exact use-case for discussify: the discussion CRDT contains cids for the comment themselves, so we need to pin those as well. After discussing this with @pgte, there's one possible solution for this:

Have a pining collaboration (a set alike crdt) where:

- Any peer can add a cid to be pinned
- The pinner would pickup added cids and pin them
- When done, the pinner would remove the cids from the CRDT

Considerations for the future:
- Ensure that only pinners can remove cids from the CRDT
- Protect against spam where peers add a lot of cids in short periods of time
